feat: 5.31.0 release (#499)
webiny/docs.webiny.com: src/docs/5.31.x/core-development-concepts/basics/routes-and-events.mdx

- API Gateway handler
- Raw handler
- EventBridge handler
- SQS handler
- S3 handler
- DynamoDB handler

### Why All The Different Handlers?
Well, the API Gateway handler and the Raw handler are actually only ones that are really required. Others are here to help with typings and some checks.

#### API Gateway Handler
This handler uses [@fastify/aws-lambda](https://github.com/fastify/aws-lambda-fastify) in the background to handle API Gateway events. It transforms [APIGatewayEvent](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/types/aws-lambda/trigger/api-gateway-proxy.d.ts#L60) into the request which `fastify` understands, and it is then used throughout the system.

For this handler to work it requires at least one [RoutePlugin](https://github.com/webiny/webiny-js/blob/754c402e5df49ffb4cd07f8c53574ebe5e125f04/packages/handler/src/plugins/RoutePlugin.ts) to be initialized. It always returns [LambdaResponse](https://github.com/fastify/aws-lambda-fastify/blob/master/index.d.ts#L13) defined in the `@fastify/aws-lambda` package.

#### Raw Handler

> This handler uses a nice `fastify` feature via which you can run any route you have previously defined. So basically, on initialisation of the handler we add some dummy route (something like `webiny-raw-event`) and then run it via the `.inject()` method on `fastify` instance.

Reviewer: ```suggestion
This handler uses a nice `fastify` feature that allows you to run any route you have previously defined. So basically, on initialization of the handler we add some dummy route (something like `webiny-raw-event`) and then run it via the `.inject()` method on `fastify` instance.
```

            onPost("/webiny", async (request, reply) => {
                // we can log the whole request body
                console.log(request.body);

                // and we can send some reply
                reply
                    .headers({
                        "x-route-example": "yes"
                    })
                    .send({
                        everything: {
                            ok: true
                        }
                    });
            });
        })
    ],
    http: { debug }
});
```

<Alert type="warning">
  Security checks will be done as on the <strong>/graphql</strong> endpoint. If you need customizations, feel free to add them.
</Alert>

<Alert type="danger">
  Routes can be added only when initializing the handler.
</Alert>

## Adding Event Handler

The idea behind all our predefined event handlers is to handle events which are not of API Gateway Event type.

Yes, you can catch API Gateway event like this, but there is no point since we have a handler specifically designed for that.
Of course, feel free to use what ever is available and can help you to achieve your goal.

### Example

> Let's say you created a part of code which sends out an SQS message, and you want to have a Lambda which handles that message.

Reviewer: ```suggestion
Let's say for example you wrote some code which sends out an SQS message, and you want to have a Lambda which handles that message.
```

> Good example would be if you want to run some calculation, asynchronously, from our GraphQL Lambda.

Reviewer: ```suggestion
A good example here would be if you want to run some calculation, asynchronously, from our GraphQL Lambda.
```

> You would insert an SQS Message and in turn it would trigger a Lambda which you have defined. That lambda should have code similar to this:

Reviewer: ```suggestion
You would insert an SQS Message and in turn it would trigger a Lambda which you have defined. 

That Lambda should have code similar to this:
```

```ts
import { createSQSHandler, createSQSEventHandler } from "@webiny/handler-aws";

const handler = createSQSHandler({
  plugins: [
      createSQSEventHandler(async({request, reply, event, context, lambdaContext}) => {

>           // "context" variable is the same as in our system - of course, if you loaded all the plugins as we do

Reviewer: ```suggestion
          // the "context" variable is the same as in our system - as long as you haven't changed Webiny's default plugin loader
```

          // because it is an sqs event, you know the type of the "event" variable and you can handle it from there
          const result = await someHeavyCalculation(event);
          
          // maybe store that result into the database?
          await storeResult(result);
          
          return reply.send({
              ok: true
          });
      })
  ]
})
```

<Alert type="warning">

>   Note that we do not actually check if the event really is an SQS event. It is expected that you initialize correct handler for event you are expecting.

Reviewer: ```suggestion
  Note that we do not actually check if the event really is an SQS event. It is expected that you initialize the correct handler for event you are expecting. If you add the wrong event it will not execute.
```

</Alert>

### Event Handler Response

> When handling an event, you can either return the `reply` object or something else, what ever you like.

Reviewer: ```suggestion
When handling an event, you can either return the `reply` object or something else, whatever you like.
```

Basically, when you return the `reply`, a standard [APIGatewayProxyResult](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/types/aws-lambda/trigger/api-gateway-proxy.d.ts#L135) is created out of the data, headers and cookies you sent.
When you return anything else other than the `reply`, it is returned as the result of the handler, and the Lambda itself.

> For example, you can send plain text or object to get the response of the Lambda without need to parse the `APIGatewayProxyResult`.

Reviewer: ```suggestion
For example, you can send plain text or object to get the response of the Lambda without the need to parse the `APIGatewayProxyResult`.
```

* `createHandler`
* `createEventHandler`

#### What Is the `createHandler` Method?
This method creates the base of Webiny system. It will initialize `fastify`, context and everything which is defined by plugins that were passed on.

We have built-in methods for different types of events:
- API Gateway
  
`createApiGatewayHandler` if imported from `@webiny/handler-aws` or `createHandler` if imported from `@webiny/handler-aws/gateway`
- Raw handler

`createRawHandler` if imported from `@webiny/handler-aws` or `createHandler` if imported from `@webiny/handler-aws/raw`
- EventBridge handler

`createEventBridgeHandler` if imported from `@webiny/handler-aws` or `createHandler` if imported from `@webiny/handler-aws/eventBridge`
- SQS handler

`createSQSHandler` if imported from `@webiny/handler-aws` or `createHandler` if imported from `@webiny/handler-aws/sqs`
- S3 handler

`createS3Handler` if imported from `@webiny/handler-aws` or `createHandler` if imported from `@webiny/handler-aws/s3`
- DynamoDB handler

`createDynamoDBHandler` if imported from `@webiny/handler-aws` or `createHandler` if imported from `@webiny/handler-aws/dynamodb`

> Depending on your preferences, you can import which ever you like. We import methods from the root of the `@webiny/handler-aws`. 

Reviewer: ```suggestion
Depending on your need, you can import which ever is most suitable. You can import methods from the root of the `@webiny/handler-aws` package. 
```

We have built-in methods for different types of events:
- API Gateway

`createApiGatewayRoute` if imported from `@webiny/handler-aws` or `createRoute` if imported from `@webiny/handler-aws/gateway`
- Raw handler

`createRawEventHandler` if imported from `@webiny/handler-aws` or `createEventHandler` if imported from `@webiny/handler-aws/raw`
- EventBridge handler

`createEventBridgeEventHandler` if imported from `@webiny/handler-aws` or `createEventHandler` if imported from `@webiny/handler-aws/eventBridge`
- SQS handler

`createSQSEventHandler` if imported from `@webiny/handler-aws` or `createEventHandler` if imported from `@webiny/handler-aws/sqs`
- S3 handler

`createS3EventHandler` if imported from `@webiny/handler-aws` or `createEventHandler` if imported from `@webiny/handler-aws/s3`
- DynamoDB handler

`createDynamoDBEventHandler` if imported from `@webiny/handler-aws` or `createEventHandler` if imported from `@webiny/handler-aws/dynamodb`

Notice that all event handlers, except one for API Gateway, are named `createEventHandler` (or `createSomethingEventHandler` when importing from root).

> We hope that name would indicate the amount of `event handlers` you can define, so:
> - route - multiple
> - event - single

Reviewer: ```suggestion
The naming convention indicates the amount of `event handlers` you can define, so `route` is multiple whereas `event` is single.
```
